cmd/subnetcmd/upgradecmd/apply.go

```diff
@@ -0,0 +1,144 @@
+func newUpgradeApplyCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "apply [subnetName]",
+		Short: "Apply upgrade bytes onto subnet nodes",
+		Long:  `Apply generated upgrade bytes to running subnet nodes to trigger a network upgrade`,
+		RunE:  applyCmd,
+		Args:  cobra.ExactArgs(1),
+	}
+
+	cmd.Flags().BoolVar(&useConfig, "config", false, "create upgrade config for future subnet deployments (same as generate)")
+	cmd.Flags().BoolVar(&useLocal, "local", false, "apply upgrade existing `local` deployment")
+	cmd.Flags().BoolVar(&useFuji, "fuji", false, "apply upgrade existing `fuji` deployment (alias for `testnet`)")
+	cmd.Flags().BoolVar(&useFuji, "testnet", false, "apply upgrade existing `testbet` deployment (alias for `fuji`)")
+	cmd.Flags().BoolVar(&useMainnet, "mainnet", false, "apply upgrade existing `mainnet` deployment")
+
+	return cmd
+}
+
+func applyCmd(cmd *cobra.Command, args []string) error {
+	subnetName := args[0]
+
+	if !app.SubnetConfigExists(subnetName) {
+		return errors.New("subnet does not exist")
+	}
+
+	sc, err := app.LoadSidecar(subnetName)
+	if err != nil {
+		return fmt.Errorf("unable to load sidecar: %w", err)
+	}
+
+	networkToUpgrade, err := selectNetworkToUpgrade(sc)
+	if err != nil {
+		return err
+	}
+
+	switch networkToUpgrade {
+	// in this case, we just are going to generate new update bytes
+	case futureDeployment:
+		return upgradeGenerateCmd(cmd, args)
+	// update a locally running network
+	case localDeployment:
+		return applyLocalNetworkUpgrade(subnetName, sc)
+	}
+
+	return nil
+}
+
+func applyLocalNetworkUpgrade(subnetName string, sc models.Sidecar) error {
+	// For a already deployed subnet, the supported scheme is to
+	// save a snapshot, and to load the snapshot with the upgrade
+
+	// first let's check update bytes actually exist
+	netUpgradeBytes, err := upgrades.ReadUpgradeFile(subnetName, app.GetSubnetDir())
+	if err != nil {
+		if err == os.ErrNotExist {
+			ux.Logger.PrintToUser("No file with upgrade specs for the given subnet has been found")
+			ux.Logger.PrintToUser("You may need to first create it with the `avalanche subnet update generate` command or import it")
+			ux.Logger.PrintToUser("Aborting this command. No changes applied")
+		}
+		return err
+	}
+
+	if err := validateUpgradeBytes(netUpgradeBytes); err != nil {
+		return err
+	}
+
+	cli, err := binutils.NewGRPCClient()
+	if err != nil {
+		return err
+	}
+	ctx := binutils.GetAsyncContext()
+
+	// get the blockchainID from the sidecar
+	blockchainID := sc.Networks[models.Local.String()].BlockchainID
+	if blockchainID == ids.Empty {
+		return errors.New(
+			"failed to find deployment information about this subnet in state - aborting")
+	}
+
+	// save a temporary snapshot
+	snapName := subnetName + tmpSnapshotInfix + time.Now().Format(timestampFormat)
+	app.Log.Debug("saving temporary snapshot for upgrade bytes", zap.String("snapshot-name", snapName))
+	_, err = cli.SaveSnapshot(ctx, snapName)
```

Comment:
what happen is the subnet has not been deployed yet?

Reply:
All very good questions!

For the latter question we need state - therefore again, this goes into the next iteration.

For the first point - if we just start it, but the subnet has not been deployed yet, the upgrade would fail (and it should fail). Hence we need again state...

So this iteration only works with already deployed and running subnet. I suggest to address these things in the next iteration. However, I also think we should not release before we have addressed this issues.